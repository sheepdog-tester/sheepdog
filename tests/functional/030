#!/bin/bash

# Test cluster snapshot

. ./common
for i in `seq 0 5`; do
	_start_sheep $i;
done
_wait_for_sheep 6

TMPDIR="$STORE/tmp"
rm -rf $TMPDIR

_cluster_format -c 6

_vdi_create test1 10M
$DOG vdi create test2 -c 3 10M

_random | $DOG vdi write test1
_random | $DOG vdi write test2
$DOG vdi read test1 | md5sum > $STORE/csum.11.org
$DOG vdi read test2 | md5sum > $STORE/csum.21.org
$DOG vdi snapshot test1
$DOG vdi snapshot test2
$DOG cluster snapshot save s1 $TMPDIR
$DOG cluster snapshot list $TMPDIR | _filter_date

_random | $DOG vdi write test1
_random | $DOG vdi write test2
$DOG vdi read test1 | md5sum > $STORE/csum.12.org
$DOG vdi read test2 | md5sum > $STORE/csum.22.org
$DOG vdi snapshot test1
$DOG vdi snapshot test2
$DOG cluster snapshot save s2 $TMPDIR
$DOG cluster snapshot list $TMPDIR | _filter_date

_vdi_list

_cleanup
for i in `seq 0 5`; do
	_start_sheep $i;
done
_wait_for_sheep 6

_cluster_format -c 6
$DOG cluster snapshot load s1 $TMPDIR
_vdi_list

$DOG vdi read test1 | md5sum > $STORE/csum.11.new
$DOG vdi read test2 | md5sum > $STORE/csum.21.new
diff -u $STORE/csum.11.org $STORE/csum.11.new
diff -u $STORE/csum.21.org $STORE/csum.21.new

_cleanup
for i in `seq 0 5`; do
	_start_sheep $i;
done
_wait_for_sheep 6

$DOG cluster snapshot load s2 $TMPDIR
_vdi_list

_cleanup
for i in `seq 0 5`; do
	_start_sheep $i;
done
_wait_for_sheep 6

$DOG cluster snapshot load 2 $TMPDIR
_vdi_list

$DOG vdi read -s 2 test1 | md5sum > $STORE/csum.12.new
$DOG vdi read -s 2 test2 | md5sum > $STORE/csum.22.new
diff -u $STORE/csum.12.org $STORE/csum.12.new
diff -u $STORE/csum.22.org $STORE/csum.22.new

$DOG vdi read test1 | md5sum > $STORE/csum.12.new
$DOG vdi read test2 | md5sum > $STORE/csum.22.new
diff -u $STORE/csum.12.org $STORE/csum.12.new
diff -u $STORE/csum.22.org $STORE/csum.22.new

$DOG vdi read -s 1 test1 | md5sum > $STORE/csum.11.new
$DOG vdi read -s 1 test2 | md5sum > $STORE/csum.21.new
diff -u $STORE/csum.11.org $STORE/csum.11.new
diff -u $STORE/csum.21.org $STORE/csum.21.new

$DOG cluster info -v | _filter_cluster_info
_vdi_create test3 10M
_vdi_list


# save or load specified vdi

_cleanup
for i in `seq 0 5`; do
	_start_sheep $i
done
_wait_for_sheep 6

_cluster_format

for i in `seq 0 9`; do
	$DOG vdi create test$i 10M
	_random | $DOG vdi write test$i
	$DOG vdi read test$i | md5sum > $STORE/csum.${i}3.org
	$DOG vdi snapshot test$i -s snap
done
_vdi_list

$DOG cluster snapshot save s3 $TMPDIR test0 test2 test4 test6 test8

$DOG cluster snapshot show s3 $TMPDIR

_cleanup
for i in `seq 0 5`; do
	_start_sheep $i
done
_wait_for_sheep 6

_cluster_format
$DOG cluster snapshot load s3 $TMPDIR
_vdi_list

for i in 0 2 4 6 8; do
	$DOG vdi read test$i | md5sum > $STORE/csum.${i}3.new
	diff -u $STORE/csum.${i}3.org $STORE/csum.${i}3.new
done

_cleanup
for i in `seq 0 5`; do
	_start_sheep $i
done
_wait_for_sheep 6

_cluster_format
$DOG cluster snapshot load s3 $TMPDIR test0 test4 test8
_vdi_list

for i in 0 4 8; do
	$DOG vdi read test$i | md5sum > $STORE/csum.${i}3.new
	diff -u $STORE/csum.${i}3.org $STORE/csum.${i}3.new
done


# delete vdi snapshot before saving cluster snapshot

echo "yes" | _cluster_format
$DOG vdi create test 10M
_random | $DOG vdi write test
$DOG vdi snapshot -s vs1 test
$DOG cluster snapshot save s4 $TMPDIR
$DOG vdi delete -s 1 test
$DOG cluster snapshot save s5 $TMPDIR
$DOG vdi snapshot -s vs2 test
$DOG cluster snapshot save s5 $TMPDIR
echo "yes" | $DOG cluster snapshot load s5 $TMPDIR
_vdi_list
echo "yes" | $DOG cluster snapshot load s5 $TMPDIR test
_vdi_list
