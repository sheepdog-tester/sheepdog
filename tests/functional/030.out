QA output created by 030
using backend plain store
Index		Tag		Snapshot Time
1		s1		DATE
Index		Tag		Snapshot Time
1		s1		DATE
2		s2		DATE
  Name        Id    Size    Used  Shared    Creation time   VDI id  Copies  Tag   Block Size Shift
s test1        1   10 MB   12 MB  0.0 MB DATE   fd32fc      6                22
s test1        2   10 MB   12 MB  0.0 MB DATE   fd32fd      6                22
  test1        0   10 MB  0.0 MB   12 MB DATE   fd32fe      6                22
s test2        1   10 MB   12 MB  0.0 MB DATE   fd3815      3                22
s test2        2   10 MB   12 MB  0.0 MB DATE   fd3816      3                22
  test2        0   10 MB  0.0 MB   12 MB DATE   fd3817      3                22
using backend plain store
using backend plain store
  Name        Id    Size    Used  Shared    Creation time   VDI id  Copies  Tag   Block Size Shift
s test1        1   10 MB   12 MB  0.0 MB DATE   fd32fc      6                22
  test1        0   10 MB  0.0 MB   12 MB DATE   fd32fd      6                22
s test2        1   10 MB   12 MB  0.0 MB DATE   fd3815      3                22
  test2        0   10 MB  0.0 MB   12 MB DATE   fd3816      3                22
using backend plain store
  Name        Id    Size    Used  Shared    Creation time   VDI id  Copies  Tag   Block Size Shift
s test1        1   10 MB   12 MB  0.0 MB DATE   fd32fc      6                22
s test1        2   10 MB   12 MB  0.0 MB DATE   fd32fd      6                22
  test1        0   10 MB  0.0 MB   12 MB DATE   fd32fe      6                22
s test2        1   10 MB   12 MB  0.0 MB DATE   fd3815      3                22
s test2        2   10 MB   12 MB  0.0 MB DATE   fd3816      3                22
  test2        0   10 MB  0.0 MB   12 MB DATE   fd3817      3                22
using backend plain store
  Name        Id    Size    Used  Shared    Creation time   VDI id  Copies  Tag   Block Size Shift
s test1        1   10 MB   12 MB  0.0 MB DATE   fd32fc      6                22
s test1        2   10 MB   12 MB  0.0 MB DATE   fd32fd      6                22
  test1        0   10 MB  0.0 MB   12 MB DATE   fd32fe      6                22
s test2        1   10 MB   12 MB  0.0 MB DATE   fd3815      3                22
s test2        2   10 MB   12 MB  0.0 MB DATE   fd3816      3                22
  test2        0   10 MB  0.0 MB   12 MB DATE   fd3817      3                22
Cluster status: running, auto-recovery enabled
Cluster store: plain with 6 redundancy policy
Cluster vnodes strategy: auto
Cluster vnode mode: node
Cluster created at DATE

Epoch Time           Version [Host:Port:V-Nodes,,,]
DATE      1 [127.0.0.1:7000:128, 127.0.0.1:7001:128, 127.0.0.1:7002:128, 127.0.0.1:7003:128, 127.0.0.1:7004:128, 127.0.0.1:7005:128]
  Name        Id    Size    Used  Shared    Creation time   VDI id  Copies  Tag   Block Size Shift
s test1        1   10 MB   12 MB  0.0 MB DATE   fd32fc      6                22
s test1        2   10 MB   12 MB  0.0 MB DATE   fd32fd      6                22
  test1        0   10 MB  0.0 MB   12 MB DATE   fd32fe      6                22
  test3        0   10 MB  0.0 MB  0.0 MB DATE   fd3662      6                22
s test2        1   10 MB   12 MB  0.0 MB DATE   fd3815      3                22
s test2        2   10 MB   12 MB  0.0 MB DATE   fd3816      3                22
  test2        0   10 MB  0.0 MB   12 MB DATE   fd3817      3                22
using backend plain store
  Name        Id    Size    Used  Shared    Creation time   VDI id  Copies  Tag   Block Size Shift
s test5        1   10 MB   12 MB  0.0 MB DATE   fd2c30      3          snap  22
  test5        0   10 MB  0.0 MB   12 MB DATE   fd2c31      3                22
s test4        1   10 MB   12 MB  0.0 MB DATE   fd2de3      3          snap  22
  test4        0   10 MB  0.0 MB   12 MB DATE   fd2de4      3                22
s test7        1   10 MB   12 MB  0.0 MB DATE   fd2f96      3          snap  22
  test7        0   10 MB  0.0 MB   12 MB DATE   fd2f97      3                22
s test6        1   10 MB   12 MB  0.0 MB DATE   fd3149      3          snap  22
  test6        0   10 MB  0.0 MB   12 MB DATE   fd314a      3                22
s test1        1   10 MB   12 MB  0.0 MB DATE   fd32fc      3          snap  22
  test1        0   10 MB  0.0 MB   12 MB DATE   fd32fd      3                22
s test0        1   10 MB   12 MB  0.0 MB DATE   fd34af      3          snap  22
  test0        0   10 MB  0.0 MB   12 MB DATE   fd34b0      3                22
s test3        1   10 MB   12 MB  0.0 MB DATE   fd3662      3          snap  22
  test3        0   10 MB  0.0 MB   12 MB DATE   fd3663      3                22
s test2        1   10 MB   12 MB  0.0 MB DATE   fd3815      3          snap  22
  test2        0   10 MB  0.0 MB   12 MB DATE   fd3816      3                22
s test9        1   10 MB   12 MB  0.0 MB DATE   fd4094      3          snap  22
  test9        0   10 MB  0.0 MB   12 MB DATE   fd4095      3                22
s test8        1   10 MB   12 MB  0.0 MB DATE   fd4247      3          snap  22
  test8        0   10 MB  0.0 MB   12 MB DATE   fd4248      3                22
1. VDI id: fd2de3, name: test4, tag: snap
2. VDI id: fd3149, name: test6, tag: snap
3. VDI id: fd34af, name: test0, tag: snap
4. VDI id: fd3815, name: test2, tag: snap
5. VDI id: fd4247, name: test8, tag: snap
using backend plain store
using backend plain store
  Name        Id    Size    Used  Shared    Creation time   VDI id  Copies  Tag   Block Size Shift
s test4        1   10 MB   12 MB  0.0 MB DATE   fd2de3      3          snap  22
  test4        0   10 MB  0.0 MB   12 MB DATE   fd2de4      3                22
s test6        1   10 MB   12 MB  0.0 MB DATE   fd3149      3          snap  22
  test6        0   10 MB  0.0 MB   12 MB DATE   fd314a      3                22
s test0        1   10 MB   12 MB  0.0 MB DATE   fd34af      3          snap  22
  test0        0   10 MB  0.0 MB   12 MB DATE   fd34b0      3                22
s test2        1   10 MB   12 MB  0.0 MB DATE   fd3815      3          snap  22
  test2        0   10 MB  0.0 MB   12 MB DATE   fd3816      3                22
s test8        1   10 MB   12 MB  0.0 MB DATE   fd4247      3          snap  22
  test8        0   10 MB  0.0 MB   12 MB DATE   fd4248      3                22
using backend plain store
using backend plain store
  Name        Id    Size    Used  Shared    Creation time   VDI id  Copies  Tag   Block Size Shift
s test4        1   10 MB   12 MB  0.0 MB DATE   fd2de3      3          snap  22
  test4        0   10 MB  0.0 MB   12 MB DATE   fd2de4      3                22
s test0        1   10 MB   12 MB  0.0 MB DATE   fd34af      3          snap  22
  test0        0   10 MB  0.0 MB   12 MB DATE   fd34b0      3                22
s test8        1   10 MB   12 MB  0.0 MB DATE   fd4247      3          snap  22
  test8        0   10 MB  0.0 MB   12 MB DATE   fd4248      3                22
    __
   ()'`;
   /\|`
  /  |   Caution! The cluster is not empty.
(/_)_|_  Are you sure you want to continue? [yes/no]: using backend plain store
Cannot execute. It may be caused by:
1. The cluster is empty.
2. All VDIs of the cluster don't have snapshots.
Fail to save snapshot to path: /tmp/sheepdog/030/tmp.
    __
   ()'`;
   /\|`
  /  |   Caution! The cluster is not empty.
(/_)_|_  Are you sure you want to continue? [yes/no]: using backend plain store
  Name        Id    Size    Used  Shared    Creation time   VDI id  Copies  Tag   Block Size Shift
s test         2   10 MB  0.0 MB   12 MB DATE   7c2b26      3           vs2  22
  test         0   10 MB  0.0 MB   12 MB DATE   7c2b27      3                22
    __
   ()'`;
   /\|`
  /  |   Caution! The cluster is not empty.
(/_)_|_  Are you sure you want to continue? [yes/no]: using backend plain store
  Name        Id    Size    Used  Shared    Creation time   VDI id  Copies  Tag   Block Size Shift
s test         2   10 MB  0.0 MB   12 MB DATE   7c2b26      3           vs2  22
  test         0   10 MB  0.0 MB   12 MB DATE   7c2b27      3                22
